src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs

```diff
@@ -103,6 +103,42 @@ public interface IArgumentCompleter
             string wordToComplete,
             CommandAst commandAst,
             IDictionary fakeBoundParameters);
+    } 
+
+    /// <summary>
+    /// A type specified by the <see cref="ArgumentCompleterAttribute"/> can implement this interface.
+    /// </summary>
+#nullable enable
+    public interface IArgumentCompleter2 : IArgumentCompleter
+    {
+        /// <summary>
+        /// Implementations of this function are called by PowerShell to complete arguments.
+        /// </summary>
+        /// <param name="commandName">The name of the command that needs argument completion.</param>
+        /// <param name="parameterName">The name of the parameter that needs argument completion.</param>
+        /// <param name="wordToComplete">The (possibly empty) word being completed.</param>
+        /// <param name="commandAst">The command ast in case it is needed for completion.</param>
+        /// <param name="fakeBoundParameters">
+        /// This parameter is similar to $PSBoundParameters, except that sometimes PowerShell cannot or
+        /// will not attempt to evaluate an argument, in which case you may need to use <paramref name="commandAst"/>.
+        /// </param>
+        /// <param name="cursorPosition">an <see cref="IScriptPosition"/>with the cursor position.</param>
+        /// <param name="replacementIndex">the index where the replacement should start.</param>
+        /// <param name="replacementLength">the length of the replacement.</param>
+        /// <returns>
+        /// A collection of completion results, most like with <see cref="CompletionResult.ResultType"/> set to
+        /// <see cref="CompletionResultType.ParameterValue"/>.
+        /// </returns>
+        IEnumerable<CompletionResult> CompleteArgument(
+            string commandName,
+            string parameterName,
+            string wordToComplete,
+            CommandAst commandAst,
+            IDictionary fakeBoundParameters,
+            IScriptPosition cursorPosition,
+            out int replacementIndex,
+            out int replacementLength
```

Comment:
Also if everything new is in one class, that makes a potentially easy way to surface it to script based completers via an automatic variable.

Reply:
Good point!

I have also considered doing the same thing with the result that PSReadline uses, so it can check for the existence of an interface, but also work with older versions.